0.19

chuwy opened the pull request:
This is based on https://github.com/http4s/rho/pull/236 and leaves @Igosuki's work almost untouched, but:

* bumps http4s to 0.19.0 (which is [still not final](https://github.com/http4s/http4s/pull/2153), I guess?)
* rebased on top of current master
* makes all tests green
* Bumps SBT to latest version
* Makes json4s a test-only dependency for swagger

#### TODO

- [x] Fix `HListToFuncSpec`
- [x] Fix docstring in `AuthedContext`
- [x] Rename few more "service" to "routes"
- [x] Add a comment in `AuthedContext` that middleware is mandatory
- [x] Should we rename `RhoService` to `RhoRoutes`?
- [ ] Bump to final 0.20.0 when available

zarthross commented on swagger/src/main/scala/org/http4s/rho/swagger/SwaggerModelsBuilder.scala:
```diff
@@ -228,8 +228,8 @@ private[swagger] class SwaggerModelsBuilder(formats: SwaggerFormats) {
-      consumes    = rr.validMedia.toList.map(_.renderString),
-      produces    = rr.responseEncodings.toList.map(_.renderString),
+      consumes    = rr.validMedia.toList.map(_.toString),
+      produces    = rr.responseEncodings.toList.map(_.toString),
```
I feel like the old `renderString` function was not the same as `toString` on a MediaRange. (Possibly not non an responseEncoding`)... `toString`'s results have a `MediaRange` and some parens...

zarthross replied:
I'm looking at this diff in http4s to see the change: https://github.com/http4s/http4s/commit/f2ca17540cde57d0e5664ac5a9e9db37b92612c1

zarthross commented on swagger/src/test/scala/org/http4s/rho/swagger/Arbitraries.scala:
```diff
@@ -62,11 +62,11 @@ object Arbitraries {
+      audio.midi,
+      image.gif,
+      text.html
+    ).map(_.toString)
```
Again, not sure `toString` should be the substitute for `renderString` on a `MediaRange`

chuwy replied:
It seems that we have `Show` instance for `MediaType`. I believe it will be the correct substitution.

zarthross replied:
I believe you are correct.

chuwy commented on examples/src/main/scala/com/http4s/rho/hal/plus/swagger/demo/Main.scala:
```diff
@@ -18,16 +15,15 @@ object Main extends StreamApp[IO] {
       .mountService(routes.staticContent combineK routes.dynamicContent, "")
       .bindLocal(port)
-      .serve
+      .serve.compile.toList.map(_.head)
```
Right, thanks. We don't want `toList` for sure. But I belive `compile.drain.as(ExitCode.Success)` is a preferred option for streaming applications inside `IOApp`.

chuwy commented:
Thank you very much for the feedback, @zarthross. I'm still a little bit puzzled by `HListToFuncSpec` as it cannot resolver `EntityEncoder[IO, IO[String]]`, but probably I'm missing something super obvious as it is quite late in where I am - I'll try to fix it tomorrow.

chuwy commented:
Aha! So, the problem is that indeed `EntityDecoder[F, F[A]]` is missing since [this commit](https://github.com/http4s/http4s/commit/13527d6c320efd25800680a8fd841365dec7965e#diff-2f7b4e737e279d203ad95a9733e8cc94). Wondering what was the reasoning.

zarthross commented on core/src/test/scala/org/http4s/rho/CodecRouterSpec.scala:
```diff
@@ -14,7 +14,7 @@ class CodecRouterSpec extends Specification {
   "A CodecRouter in a RhoService" should {
 
-    val service = new RhoService[IO] {
+    val routes = new RhoService[IO] {
```
This is still a `RhoService`, so i don't know that the rename is appropriate here.

chuwy replied:
But we're immediately converting it to `HttpRoutes[IO]` below, so I think all these are justified.

zarthross replied:
¯\_(ツ)_/¯ 
Not really that big of deal.  Not a deal breaker for me either way.

zarthross replied:
(Removed other similar comments for brevity.)

zarthross commented on core/src/test/scala/org/http4s/rho/ParamDefaultValueSpec.scala:
```diff
@@ -6,340 +6,340 @@ import org.specs2.mutable.Specification
 
   "GET /test1" should {
-    val service = new RhoService[IO] {
+    val routes = new RhoService[IO] {
```
This is still a `RhoService`, so i don't know that the rename is appropriate here.

zarthross commented on core/src/test/scala/org/http4s/rho/ParamDefaultValueSpec.scala:
```diff
@@ -6,340 +6,340 @@ import org.specs2.mutable.Specification
 
   "GET /test2" should {
-    val service = new RhoService[IO] {
+    val routes = new RhoService[IO] {
```
This is still a `RhoService`, so i don't know that the rename is appropriate here.

zarthross commented on core/src/test/scala/org/http4s/rho/ParamDefaultValueSpec.scala:
```diff
@@ -6,340 +6,340 @@ import org.specs2.mutable.Specification
 
   "GET /test3" should {
-    val service = new RhoService[IO] {
+    val routes = new RhoService[IO] {
```
as above

zarthross commented on core/src/test/scala/org/http4s/rho/ParamDefaultValueSpec.scala:
```diff
@@ -6,340 +6,340 @@ import org.specs2.mutable.Specification
 
   "GET /test4" should {
-    val service = new RhoService[IO] {
+    val routes = new RhoService[IO] {
```
as above

zarthross commented on core/src/test/scala/org/http4s/rho/ParamDefaultValueSpec.scala:
```diff
@@ -6,340 +6,340 @@ import org.specs2.mutable.Specification
 
   "GET /test5" should {
-    val service = new RhoService[IO] {
+    val routes = new RhoService[IO] {
```
as above

zarthross commented on core/src/test/scala/org/http4s/rho/ParamDefaultValueSpec.scala:
```diff
@@ -6,340 +6,340 @@ import org.specs2.mutable.Specification
 
   "GET /test6" should {
-    val service = new RhoService[IO] {
+    val routes = new RhoService[IO] {
```
as above

zarthross commented on core/src/test/scala/org/http4s/rho/ParamDefaultValueSpec.scala:
```diff
@@ -6,340 +6,340 @@ import org.specs2.mutable.Specification
   "GET /test14" should {
 
-    val service = new RhoService[IO] {
+    val routes = new RhoService[IO] {
```
Gave up marking the all...

zarthross commented on core/src/main/scala/org/http4s/rho/CompileService.scala:
```diff
@@ -42,7 +42,7 @@ object CompileService {
     * @return An `HttpService`
     */
-  def foldServices[F[_]: Monad](routes: Seq[RhoRoute.Tpe[F]], filter: RhoMiddleware[F]): HttpService[F] = {
+  def foldServices[F[_]: Monad](routes: Seq[RhoRoute.Tpe[F]], filter: RhoMiddleware[F]): HttpRoutes[F] = {
```
Hmm... TBH, I think this function needs a bit more than a rename, but it might be out of the scope of this PR.  So, yes i think you should rename it to `foldRoutes`.  

If you have time, I would say remove the `filter` parameter and have the [`ServiceBuilder.toService`](https://github.com/http4s/rho/blob/v0.18.0/core/src/main/scala/org/http4s/rho/ServiceBuilder.scala#L18) function which calls `foldServices` run the middleware on the routes before passing them in.  This reduces the complexity of this function, and makes the name make more sense.

rossabaker commented:
Yeah, sorry about 0.19.0 not being final.  I messed up the tagging.

Would you like to do a milestone release on this so people can try it?

zarthross commented:
I think there is a few more changes @chuwy would like to make before we merge/release, but once he finishes those check boxes, I would like to get this released.  I have some projects I would like to get a head start updating before the 0.19.1 release.

chuwy commented:
@rossabaker, @zarthross I made some more "service -> routes" renamings to be consistent with core and [updated AuthedContext](https://github.com/http4s/rho/pull/253/commits/ec2ba9e5e90656cd58d9aa5821ffec757141e26d) (I believe unsafe `getAuth` was a bug - I stumbled upon it when first used rho). It would be nice to get M1 when you have time.

Last question I have is whether we should rename `RhoService` to `RhoRoutes`. I vote for yes.

zarthross commented on core/src/main/scala/org/http4s/rho/ServiceBuilder.scala:
```diff
@@ -7,15 +7,15 @@ import shapeless.HList
     * @param filter [[RhoMiddleware]] to apply to the collection of routes.
     * @return An `HttpService` which can be mounted by http4s servers.
     */
   def toService(filter: RhoMiddleware[F] = identity): HttpRoutes[F] =
```
`ServiceBuilder` and its `toService` function also need a rename.

chuwy replied:
Agree, but I think we should handle this along with `RhoService`

zarthross replied:
Yeah, I agree with the `RhoService` rename.  It makes sense in context with what http4s is trying to do.

chuwy replied:
My worry is similar: we're getting too many "route" in different places that clash with each other and core's. We're ending up with `RouteFoo.toRoutes(routes).asRoutes` roughly speaking. But still I believe this is a right choice.

zarthross commented on core/src/main/scala/org/http4s/rho/AuthedContext.scala:
```diff
@@ -57,13 +48,13 @@ class AuthedContext[F[_]: Monad, U] extends FailureResponseOps[F] {
-  }
+  /** Get the authInfo object from request */
+  def getAuth(req: Request[F]): Option[U] =
+    req.attributes.get(authKey)
```
Good catch, `get` should almost never be called on an `Option`, however, there is a more fundamental problem here that is beyond the scope of this PR to fix.   You should not be able to call `getAuth` if your `RhoService` hasn't been wrapped in a `AuthMiddleware` but the current design allows this.  So, for now, I propose we just make the error message you get for calling `getAuth` better, and file an issue to correct our auth framework so it doesn't allow this.

Thinking replacing `get` with `getOrElse(throw new RuntimeException("Misconfiguration: RhoService called getAuth but has not had AuthMiddleware applied"))`.  Certainly open to other exception messages, but looking for something more descriptive than `get` would throw alone.

chuwy replied:
Raised it [here](https://github.com/http4s/rho/issues/256). For me `None` makes much more sense than `RuntimeException` because design of `AuthMiddleware` implies that it can return `None`.

zarthross commented on core/src/test/scala/org/http4s/rho/AuthedContextSpec.scala:
```diff
@@ -26,13 +26,22 @@ object MyAuth extends AuthedContext[IO, User]
+  GET / "private" / 'place |>> { (req: Request[IO], path: String) =>
+    getAuth(req) match {
+      case Some(user) => Ok(s"${user.name} at $path")
+      case None => Forbidden(s"not authenticated at $path")
```
Unfortunately, if `getAuth` returns `None` it doesn't mean the user was not authenticated, it means the `AuthMiddleware` wasn't applied.

chuwy replied:
Our `authenticated` middleware always returns `Some`, but in case where it returns `None` (which is still valid) - it will be indistinguishable from not applied middleware. So I have we have plenty of slightly broken APIs, but this one from my POV is less broken. We just need to mention in comments that `AuthMiddleware` always should be applied. Until we come up with best strategy.

zarthross commented:
@chuwy How many of those checkboxes do you want to complete before this gets merged in?

chuwy commented:
@zarthross I think two at least. Should be done in next couple of days.

zarthross commented:
I think it would probably be wise to go through with the `RhoService` to `RhoRoutes` rename to follow suite with the Http4s rename.

chuwy commented:
Sure. Totally agree. I meant "all except Bump to final 0.19.1", so this renaming will be included for sure.

chuwy commented:
Also, I assume you're fine with `Option` in `AuthedContext`. I believe that it is a not an ideal solution, but still better than runtime exception.

ChristopherDavenport commented on project/Dependencies.scala:
```diff
@@ -2,7 +2,7 @@ import sbt._
 
 object Dependencies {
-  lazy val http4sVersion = "0.18.19"
+  lazy val http4sVersion = "0.19.0"
```
The currently version is `0.19.0-M4`, 0.19.0 was mistagged and should be avoided if possible.

chuwy commented:
Hey @zarthross, sorry for delay. This has been updated to http4s 0.20.0-M1. I also believe I addressed all your feedback and everything I wanted to change.

I assume we're not pursuing version-parity with http4s, so I left it as 0.19, not 0.20.

If everything is correct, it should be ready for M1.

rossabaker commented:
Version parity has always been coincidental, but maybe it's nice since it has recently been in lockstep.  I guess it might get more complicated when both projects hit 1.0. I don't have a strong opinion there.
